cloudinit/sources/__init__.py

```diff
@@ -221,6 +223,16 @@ def __init__(self, sys_cfg, distro, paths, ud_proc=None):
     def __str__(self):
         return type_utils.obj_name(self)
 
+    def _unpickle(self, ci_pkl_version: int) -> None:
+        """Perform deserialization fixes for Distro."""
+        if "run_dir" not in self.paths.__dict__:
```

Comment:
```suggestion
        if not hasattr(self.paths, "run_dir"):
```

because it handles properties appropriately:

```py
In [1]: class Foo:
   ...:     @property
   ...:     def run_dir(self):
   ...:         pass
   ...: 

In [2]: "run_dir" in Foo().__dict__
Out[2]: False

In [3]: hasattr(Foo(), "run_dir")
Out[3]: True
```